File: dailyques/Q799.py
class Solution(object):
    def champagneTower(self, poured, query_row, query_glass):
        dp = [0.0] * 102
        dp[0] = poured

        for i in range(query_row):
            nxt = [0.0] * 102
            for j in range(i + 1):
                if dp[j] > 1.0:
                    extra = (dp[j] - 1.0) / 2.0
                    nxt[j] += extra
                    nxt[j + 1] += extra
            dp = nxt

        if dp[query_glass] > 1.0:
            return 1.0
        return dp[query_glass]

File: dailyques/Q799.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    double champagneTower(int poured, int query_row, int query_glass) {
        vector<double> dp(102, 0.0);
        dp[0] = poured;

        for (int i = 0; i < query_row; i++) {
            vector<double> next(102, 0.0);
            for (int j = 0; j <= i; j++) {
                if (dp[j] > 1.0) {
                    double extra = (dp[j] - 1.0) / 2.0;
                    next[j] += extra;
                    next[j + 1] += extra;
                }
            }
            dp.swap(next);
        }

        return min(1.0, dp[query_glass]);
    }
};